src/Forms/MoneyField.php

```diff
@@ -86,48 +84,124 @@ public function __clone()
+            reset($allowedCurrencies);
+            $currencyValue = key($allowedCurrencies);
+        } elseif ($allowedCurrencies) {
+            // Dropdown field for multiple currencies
+            $field = DropdownField::create(
                 "{$name}[Currency]",
                 _t('MoneyField.FIELDLABELCURRENCY', 'Currency'),
-                ArrayLib::is_associative($allowedCurrencies)
-                    ? $allowedCurrencies
-                    : array_combine($allowedCurrencies, $allowedCurrencies)
+                $allowedCurrencies
             );
         } else {
-            $field = new TextField(
+            // Free-text entry for currency value
+            $field = TextField::create(
                 "{$name}[Currency]",
                 _t('MoneyField.FIELDLABELCURRENCY', 'Currency')
             );
         }
 
         $field->setReadonly($this->isReadonly());
         $field->setDisabled($this->isDisabled());
+        if ($currencyValue) {
+            $field->setValue($currencyValue);
+        }
+        $this->fieldCurrency = $field;
         return $field;
     }
 
-    public function setValue($val)
+    public function setSubmittedValue($value, $data = null)
     {
-        $this->value = $val;
+        if (empty($value)) {
+            $this->value = null;
+            $this->fieldCurrency->setValue(null);
+            $this->fieldAmount->setValue(null);
+            return $this;
+        }
 
-        if (is_array($val)) {
-            $this->fieldCurrency->setValue($val['Currency']);
-            $this->fieldAmount->setValue($val['Amount']);
-        } elseif ($val instanceof DBMoney) {
-            $this->fieldCurrency->setValue($val->getCurrency());
-            $this->fieldAmount->setValue($val->getAmount());
+        // Handle submitted array value
+        if (!is_array($value)) {
+            throw new InvalidArgumentException("Value is not submitted array");
         }
 
-        // @todo Format numbers according to current locale, incl.
-        //  decimal and thousands signs, while respecting the stored
-        //  precision in the database without truncating it during display
-        //  and subsequent save operations
+        // Update each field
+        $this->fieldCurrency->setSubmittedValue($value['Currency'], $value);
+        $this->fieldAmount->setSubmittedValue($value['Amount'], $value);
 
+        // Get data value
+        $this->value = $this->dataValue();
+        return $this;
+    }
+
+    public function setValue($value, $data = null)
+    {
+        if (empty($value)) {
+            $this->value = null;
+            $this->fieldCurrency->setValue(null);
+            $this->fieldAmount->setValue(null);
+            return $this;
+        }
+
+        // Convert string to array
+        // E.g. `44.00 NZD`
+        if (is_string($value) &&
+            preg_match('/^(?<amount>[\\d\\.]+)( (?<currency>\w{3}))?$/i', $value, $matches)
```

Comment:
Oh dear, I'm throwing an exception... I'd intended to support that array format, sorry!